rm: linux-lmp-fs: add page about lmp file structure (#613)
foundriesio/docs: source/reference-manual/linux/linux-lmp-fs.rst

.. _ref-linux-lmp-fs:

LmP File Structure
==================

> LmP uses the `OSTree`_ upgrade system to cover platform updates. It brings a modified file system structure, which details are covered in this page along with some tips about the LmP structure and OTA behavior.

Reviewer: ```suggestion
LmP uses the `OSTree`_ upgrade system to cover platform updates. It brings a modified file system structure, the details for which are covered on this page, along with tips about LmP structure and OTA behavior.
```

   total 17
   lrwxrwxrwx   2 root root    7 Jul 31 22:32 bin -> usr/bin
   drwxr-xr-x   5 root root 1024 Aug  1 15:44 boot
   drwxr-xr-x  12 root root 3340 Aug  2 17:17 dev
   drwxr-xr-x  39 root root 4096 Aug  1 15:46 etc
   lrwxrwxrwx   2 root root   17 Jul 31 22:32 home -> var/rootdirs/home
   lrwxrwxrwx   2 root root    7 Jul 31 22:32 lib -> usr/lib
   lrwxrwxrwx   2 root root   18 Jul 31 22:32 media -> var/rootdirs/media
   lrwxrwxrwx   2 root root   16 Jul 31 22:32 mnt -> var/rootdirs/mnt
   lrwxrwxrwx   2 root root   16 Jul 31 22:32 opt -> var/rootdirs/opt
   lrwxrwxrwx   2 root root   14 Jul 31 22:32 ostree -> sysroot/ostree
   dr-xr-xr-x 148 root root    0 Aug  2 17:16 proc
   lrwxrwxrwx   2 root root   12 Jul 31 22:32 root -> var/roothome
   drwxrwxrwt  20 root root  540 Aug  2 17:17 run
   lrwxrwxrwx   2 root root    8 Jul 31 22:32 sbin -> usr/sbin
   lrwxrwxrwx   2 root root   16 Jul 31 22:32 srv -> var/rootdirs/srv
   dr-xr-xr-x  12 root root    0 Aug  2 17:16 sys
   drwxr-xr-x   5 root root 4096 Jul 31 22:34 sysroot
   drwxrwxrwt   7 root root  140 Aug  2 17:17 tmp
   drwxr-xr-x  12 root root 4096 Jan  1  1970 usr
   drwxr-xr-x  13 root root 4096 Aug  1 15:46 var

> OSTree brings a **read-only** file system, where the disk partition is ``/sysroot`` and the root that is mounted on ``/`` (as above) is ``/sysroot/ostree/deploy/lmp/deploy/<sha>``.

Reviewer: ```suggestion
OSTree brings a **read-only** file system, where the disk partition is ``/sysroot`` and the root mounted at ``/`` (as above) is ``/sysroot/ostree/deploy/lmp/deploy/<sha>``.
```

> Some folders should not be manipulated in order to ensure proper behavior of OTA. The ``/usr`` path is an example of a critical folder that is mounted as read-only and should not be touched as that is covered by an LmP OTA.

Reviewer: ```suggestion
Some folders should not be manipulated in order to ensure proper behavior of OTA. The ``/usr`` path is an example of a critical folder that is mounted as read-only and should not be touched—this is covered by LmP OTA.
```

.. _ref-linux-lmp-fs:

LmP File Structure
==================

> LmP uses the `OSTree`_ upgrade system to cover platform updates. It brings a modified file system, which details are covered in this page along with some tips about the LmP structure and OTA behavior.

Reviewer 2: modified file system structure

OSTree File System
------------------

.. prompt::

>    # ls -l /

Reviewer 3: `tree -L 1` gives a better result than `ls` in my opinion

```
$ tree -L 1 /
/
├── bin -> usr/bin
├── boot
├── dev
├── efi
├── etc
├── home
├── lib -> usr/lib
├── lib64 -> usr/lib
├── mnt
├── opt
├── proc
├── root
├── run
├── sbin -> usr/bin
├── srv
├── sys
├── tmp
├── usr
└── var

```

Author: tree doesn't seem to be enabled by default in LmP, is this correct?

Reviewer 3: you are rigth! the tree is not available so drop my comment.

Persistent Storage
------------------

These folders are treated like persistent storage on the device and have particularities with regards to updates:

* ``/var``

   * Main persistent storage, writable directory.

   * Most important files are located under ``/var`` (see :ref:`ref-linux-lmp-fs-important-files`).

   * Not covered by OTA.

* ``/etc``

   * Covered by OTA if untouched.

   * As soon as OSTree detects a change in this directory, it performs a 3-way merge using the old default configuration, the active system's ``/etc``, and the new default configuration. This is then not covered by OTA anymore.

   * For the reason above, Foundries.io do not recommend setting system configurations in this directory. It is more reliable to set critical configurations under ``/usr/lib`` as that is always covered by OTA.

.. _ref-linux-lmp-fs-important-files:

Important Files and Folders
---------------------------

* ``/var/sota``: Stores critical OTA files.

.. prompt::

>    # ls -l /var/sota/

Reviewer 3: same here for tree

   HOME_URL="https://foundries.io/"
   SUPPORT_URL="https://support.foundries.io/"
   DEFAULT_HOSTNAME="qemuarm64-secureboot"
   LMP_MACHINE="qemuarm64-secureboot"
   LMP_FACTORY="<factory-name>"
   LMP_FACTORY_TAG="<tag>"
   IMAGE_ID=lmp-factory-image
   IMAGE_VERSION=<os-target>

Tips and Suggestions
--------------------

* A :ref:`systemd service <ref-troubleshooting_systemd-service>` can be used in case a file in a directory not covered by OTA needs to be updated.

* It is recommended to store custom user files under ``/var/local``. Keep any custom files location in mind when implementing a :ref:`ref-factory-device-reset`.

* The full initial Target information (includes containers and LmP) just after the provisioning of a device can be checked with:

.. prompt::

   # cat /var/sota/import/installed_versions | grep "\"version\""
      "version": "102",

After the first OTA, it can be read in ``/var/sota/current-target`` as discussed previously.

* If enabling :ref:`ref-linux-persistent-log`, ``/var/log`` is used to store system logs. It is recommended to mount it in an additional storage so it does not take valuable internal storage space which could impact the OTA behavior.

> * A testing hack is to remount the ``/usr`` folder as **read-write** for testing purposes:

Reviewer 2: I question if we should really expose this hack, as it might bite us later on.

Reviewer 2: What we can probably say here is that /usr/local is also rw (link to /var) and it can be used to replace system related components if needed by the user (during development). /usr/local is also not managed by ostree.

Author: I'll drop it since we only do it when debugging with the customer.
